pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -5513,102 +5565,149 @@ def StorageSiteWaterQualityRule(b, s, w, t):
     )
     # endregion
 
     # region Treatment
-    def TreatmentWaterQualityRule(b, r, w, t):
+    def TreatmentFeedWaterQualityRule(b, r, qc, t):
         constraint = (
             sum(
-                b.parent_block().v_F_Piped[n, r, t] * b.v_Q[n, w, t]
+                b.parent_block().v_F_Piped[n, r, t] * b.v_Q[n, qc, t]
                 for n in b.parent_block().s_N
                 if b.parent_block().p_NRA[n, r]
             )
             + sum(
-                b.parent_block().v_F_Piped[s, r, t] * b.v_Q[s, w, t]
+                b.parent_block().v_F_Piped[s, r, t] * b.v_Q[s, qc, t]
                 for s in b.parent_block().s_S
                 if b.parent_block().p_SRA[s, r]
             )
             + sum(
-                b.parent_block().v_F_Trucked[p, r, t] * b.p_nu_pad[p, w]
+                b.parent_block().v_F_Trucked[p, r, t] * b.p_nu_pad[p, qc]
                 for p in b.parent_block().s_PP
                 if b.parent_block().p_PRT[p, r]
             )
             + sum(
-                b.parent_block().v_F_Trucked[p, r, t] * b.p_nu_pad[p, w]
+                b.parent_block().v_F_Trucked[p, r, t] * b.p_nu_pad[p, qc]
                 for p in b.parent_block().s_CP
                 if b.parent_block().p_CRT[p, r]
             )
-        ) == b.v_Q[r, w, t] * (
-            b.parent_block().v_F_ResidualWater[r, t]
-            + b.parent_block().v_F_TreatedWater[r, t]
+        ) == b.v_Q[r, qc, t] * b.parent_block().v_F_TreatmentFeed[r, t]
+
+        return process_constraint(constraint)
+
+    model.quality.TreatmentFeedWaterQuality = Constraint(
+        model.s_R,
+        model.s_QC,
+        model.s_T,
+        rule=TreatmentFeedWaterQualityRule,
+        doc="Treatment Feed water quality",
+    )
+
+    def TreatmentWaterQualityRule(b, r, qc, t):
+        constraint = (
+            b.v_Q[r, qc, t] * b.parent_block().v_F_TreatmentFeed[r, t]
+            == b.v_Q[r + treated_water_label, qc, t]
+            * b.parent_block().v_F_TreatedWater[r, t]
+            + b.v_Q[r + residual_water_label, qc, t]
+            * b.parent_block().v_F_ResidualWater[r, t]
         )
+
         return process_constraint(constraint)
 
     model.quality.TreatmentWaterQuality = Constraint(
         model.s_R,
-        model.s_W,
+        model.s_QC,
         model.s_T,
         rule=TreatmentWaterQualityRule,
         doc="Treatment water quality",
     )
-    # Water quality of water that has been treated
-    # NOTE: Water quality changes by treatment technologies is not currently modeled
-    def TreatedWaterWaterQualityRule(b, r, w, t):
-        constraint = (b.parent_block().v_F_TreatedWater[r, t]) == b.v_Q[
-            r + treatment_intermediate_label, w, t
-        ] * (
-            sum(
-                b.parent_block().v_F_Piped[r, p, t]
-                for p in b.parent_block().s_CP
-                if b.parent_block().p_RCA[r, p]
+
+    def TreatedWaterQualityLHSRule(b, r, wt, qc, t):
```

Comment:
I see that you have rewritten this constraint to use a big-M formulation, but it doesn't seem like the documentation has been updated to reflect this change in the model. Can you please update the documentation so that it shows the new form of the constraints?

Reply:
Agreed that the documentation needs updating for these new constraints / adjustment to WaterQualityRule. Updates needed in the general model and water quality sections.

Reply:
Updated!

Reply:
Thank you for raising a valid point about the efficiency of using an expr for the model.p_M_Flow_Conc parameter. To address your concern and avoid any confusion with the bilinear term unit, I have defined model.p_M_Flow_Conc as a separate parameter that is initialized using model.p_M_Flow and model.p_M_Concentration. This approach also allows us to keep model.p_M_Concentration as the big-M parameter for concentration and use it in our MINLP.